### DISCUSSION TITLE
Server performance metrics collection

### DISCUSSION THREAD

**Discussion by passing** [jamulussoftware/jamulus#1770] [General]
Mainly out of curiosity, I collected system resource usage metrics of my Jamulus Server along with the number of Clients connected during a session with up to 13 users today. The result looks quite interesting:

![image](https://user-images.githubusercontent.com/8142186/119903771-7c11ef00-bf49-11eb-8295-1de07670165d.png)

The lower panel shows the CPU divided by number of clients, and that doesn't get bigger for more clients (at least for up to 13).
I assume the reason is the audio decoding and encoding create the most of the CPU usage, as that should be proportional to the number of clients

To be noted is that these were real users with different / random audio quality settings.

I put the configuration for collecting these metrics (using telegraf) here https://github.com/passing/jamulus-telegraf/
(I used Influx for storing the data and Grafana for the dashboard)

**Comment by ann0see** [comment]
This looks interesting! 

So we should focus on improving the encoding/decoding functions?

**Reply by passing** [reply]
> This looks interesting!
> 
> So we should focus on improving the encoding/decoding functions?

I think from the data I collected so far, it gives some first indication but no reliable answer to that. It may even be misleading as I'd assume clients using a lower audio quality create less CPU load.
It would probably make sense to collect metrics on a public server for a longer time to get more/better data to analyze, so differences from different audio settings of clients average out. If there's interest in doing that I'd be happy to support if needed.

I'm not sure if there's much room for improving the encoding/decoding functions. If that's mainly on the OPUS codec you could probably just reduce the audio quality. But could also be a "feature" to be able to limit that on a server to be able to handle more clients.

Apart from that it may be that with a very high number of clients the CPU usage will eventually start growing more due to the mixing of audio streams

**Comment by gene96817** [comment]
Very interesting!
How much CPU did you give to the server? What is your clock rate?
It will take me some time to learn how to get it installed on my server. 
This promises to help answer how much processing is needed for different numbers of participants.

**Reply by passing** [reply]
The server I collected the data on, is an AWS t3.nano instance with 2 vCPUs / 3.1 GHz clock speed.
Having seen about 50% of the CPUs used with ~20 clients in the past, I have set the limit on that server to 30 clients to still have some headroom.

**Comment by hoffie** [comment]
Cool! :)
Do you plan to extend this further? I'm asking because I've got something similar in mind, have a draft regarding implementation laying around (and don't get to posting it :/) and started running some instrumentation tests.

Maybe we can unite the efforts. :)

**Reply by passing** [reply]
I did not really plan anything further, yet, but if you have some specific ideas, I'd be happy to help!
